### PR TITLE
Phase-based operations

### PR DESCRIPTION
**Pull request by dasanchez**
- Added a Delegator class
- Added a Validator class
- delegator and validator replacements are now done with string replacements (via `sed`)
- Categorized operations into string-replacement (sed) and json ones
- User must submit a set of tasks that conform to the right order (string replacement first, followed by json operations)
- Added tests with pytest

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,38 +25,42 @@ Full documentation for the module can be accessed with `pydoc cosmos_genesis_tin
...
+
+new_del = Delegator()
+new_del.address = 'cosmos456'
+new_del.public_key = 'defuvw'
+
+tinker.add_task(tinker.replace_delegator,
+    old_delegator=old_del,
+    new_delegator=new_del)
+    
+tinker.add_task(tinker.increase_balance,
```
Comment (uditvira) [inline README.md:60]: love this api

File: cosmos_genesis_tinker.py [hunk 2]
```diff
@@ -205,19 +488,43 @@ def generate_shasum(self):
         content_bytes = content.encode('utf-8')
         return sha256(content_bytes).hexdigest()
 
-    def save_file(self, path):
+    def get_bonded_pool_address(self):
         """
-        Save a modified genesis file back to JSON
+        Look through .app_state.auth.accounts,
+        check for "name": "bonded_tokens_pool"
```
Comment (uditvira) [inline cosmos_genesis_tinker.py:501]: nice! this is much better than hardcoding

File: legacy/genesis_theta_local.py [hunk 3]
```diff
@@ -61,19 +61,25 @@
...
+LOCAL_CONS_ADDRESS = "cosmosvalcons16k44u3v0m8uevnh4p2qy2xm08y3wdf92xsc3ve"
 
 PREPROCESSED_FILE = GENESIS_ARCHIVE + ".preprocessed.json"
 
 shutil.copy2(GENESIS_ARCHIVE, PREPROCESSED_FILE)
 
-subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_SELF_DELEGATION_ADDR    + "%" + LOCAL_SELF_DELEGATION_ADDR   + "%g",PREPROCESSED_FILE])
-subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_SELF_DELEGATION_PUBKEY  + "%" + LOCAL_SELF_DELEGATION_PUBKEY + "%g",PREPROCESSED_FILE])
-subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_VALIDATOR_OP_ADDRESS    + "%" + LOCAL_VALIDATOR_OP_ADDRESS   + "%g",PREPROCESSED_FILE])
+subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_SELF_DELEGATION_ADDR +
```
Comment (uditvira) [inline legacy/genesis_theta_local.py]: should we remove the legacy folder @dasanchez? not sure what the value in keeping it is?

File: cosmos_genesis_tinker.py [hunk 4]
```diff
@@ -1,108 +1,385 @@
...
+        Run the list of tasks:
+        - All byte operations are done before the json ones
+        - All byte operations are done on a the pre-processing
+        """
+
+        if self._task_list.tasks() != self._task_list.user_tasks():
+            print('Invalid sequence: replace_validator and replace_delegator '
+                  'must come before all other functions.')
+            print('Expected order:', str([task.func.__name__ for task in self._task_list.tasks()]))
+            return True
```
Comment (uditvira) [inline cosmos_genesis_tinker.py:308]: great!

**Review by uditvira — APPROVED** [review]
This is very impressive work. Thank you @dasanchez 

You probably already did, but curious if you explicitly compared shasum from legacy version and this new version?

**Comment by dasanchez** [comment]
> curious if you explicitly compared shasum from legacy version and this new version?
@uditvira Good point, I did not check the shasum.
I will run the mainnet exported genesis through both versions and compare the shasum between the two before I merge.

File: legacy/genesis_theta_local.py (hunk 3 above)
Comment (dasanchez) [reply legacy/genesis_theta_local.py]: I'd moved the old examples to a legacy folder just to have them as reference, but yes, I'd rather delete them instead.

**Comment by dasanchez** [comment]
@uditvira 
The shasum didn't match when comparing a theta testnet genesis file generated with the old API to one generated with the new API.

The genesis file lists balances with amount first, and denom second: https://github.com/hyphacoop/cosmos-genesis-tinkerer/blob/312ef6dde0443663dde7962c3a97e033080f90cf/tests/fresh_genesis.json#L119

When I was writing the tests I noticed the `create_coin` function had those reversed (denom first, amount second): https://github.com/hyphacoop/cosmos-genesis-tinkerer/blob/59c4fc08a6bcad9600d286932ec985abd187d23b/cosmos_genesis_tinker.py#L433

So I changed this function at the time to match the genesis file order. That caused the shasum mismatch!

I have updated `create_coin` again to match the original version (denom first, amount second): https://github.com/hyphacoop/cosmos-genesis-tinkerer/blob/312ef6dde0443663dde7962c3a97e033080f90cf/cosmos_genesis_tinker.py#L627

I don't think gaiad will care if these are reversed when it's parsing the json.

I have removed the legacy folder and updated the theta genesis script to use the new API as an example for stateful genesis tinkering.

**The shasums now match**

File: legacy/genesis_theta_local.py (hunk 3 above)
Comment (dasanchez) [reply legacy/genesis_theta_local.py]: Deleted the legacy folder and turned the theta tinker script into a stateful genesis example.

**Pull request merged**